BE#42-Admin-API-Endpoints

SoySauceLabs opened the pull request:
# TOpreDOs:

- [ ] `source env_vars.txt` from **Project Documentation > Wiki+Reference > BE > env_vars.txt**

# TOtestDOs:

- [ ] Reference the `/admin` endpoints in the API document from **Project Documentation > Wiki+Reference > API >Basics**
- [ ] Ask Matt for test objects if the document is too enigmatic.

# Itinerary

- [ ] Sip it.

scotte216 commented on src/admin/views.py:
```diff
@@ -0,0 +1,517 @@
+
+    auth_token = str(data.get('auth_token', ''))
+    admin_username = str(data.get('admin_username', ''))
+    user_edit_info = data.get('edit_user_info', '')
```
user_edit_info vs edit_user_info... Seems like these should have the same name.

SoySauceLabs replied:
fixed!

scotte216 commented on src/cassy.py:
```diff
@@ -557,3 +557,553 @@ def change_user_email(username, new_email):
+            'securitytoken': new_user_info.get('securitytoken', ''),
+            'subenddate': new_user_info.get('subenddate', ''),
+            'userid': int(new_user_info.get('userid', '')),
+            'vineyards': new_user_info.get('vineyards', ''),
```
Is there a mechanism that does some kind of checking about these parameters? Like that userids are positive integers w/o duplicate (or whatever), sub end date is actually a date in the future, etc.

SoySauceLabs replied:
Good catch! There is not currently, but I'll see if I can add some reasonable checks in.

scotte216 commented on src/cassy.py:
```diff
@@ -557,3 +557,553 @@ def change_user_email(username, new_email):
+            'userid': rows[0].userid,
+            'vineyards': rows[0].vineyards,
+        }
+        if (user_edit_info.get('username', '') != ''):
```
Is there a way to do something like:
for each key in user_edit_info: 
    if user_edit_info.get(key, '') != ''):
        edit_row[key] = user_edit_info.get(key, '')

SoySauceLabs replied:
Yes, actually I believe that would work quite well here, I'll give it a touch up!

scotte216 commented:
Did a code review. Still need to do manual testing.

scotte216 commented:
endpoint admin/user/new 
Using an existing username from the DB but with a different pwd:

Expected behavior: tell us the username is bad / already exists.

Actual behavior: will accept duplicate usernames as long as the passwords are different.

scotte216 commented:
endpoint admin/user/new

Create a new user with junk for the subenddate

Expected behavior: Error about bad time/date

Actual behavior: 500 -- Unknown error occurred 

NOTE: admin/user/subscription returns a different error when the date is bad:
"unknown": "An unknown error occurred. time data '2017-o1-31' does not match format '%Y-%m-%d'"

scotte216 commented:
endpoint admin/user/new
Not sure if this is a problem but -- I'm allowed to set the subenddate to a date in the past.

SoySauceLabs commented:
All righty, I've addressed constraining the usernames to be unique, and fixed the subscription date checking for correct format and "futurama-ness" of the date! 📆

scotte216 commented:
When creating a new user with username "" I get the following error response

"sub_end_date_invalid": "Subscription end date is invalid. username_invalid"

The subscription date IS valid, but the username is invalid.

scotte216 commented:
The username is allowed to be a single space: " " 
I didn't want to make a ton of such users, but I imagine it's just reading spaces as another character. Probably want to trim whitespace.

scotte216 commented:
I'm being pickier with the user info since it seems like the kind of thing where that's most likely to be screwed up.

scotte216 commented:
/admin/vineyard
With a vineyard ID of a non-existent vineyard (20), I got the following response:
 "vineyard_bad_id": "A vineyard ID must be a positive integer."

Expected:
'The vineyard ID was not found.'

scotte216 commented:
vineyard/new
Allowed create of vineyard with vineyard "name" attribute missing.

API doc says it's required.

Also allowed creation of a new vineyard with missing boundary attribute.

scotte216 commented:
vineyard/edit

Incorrect error when using a non-existent vineyard ID:
Expected: The vineyard ID was not found
Actual: "A vineyard ID must be a positive integer."

scotte216 commented:
vineyard/

Strange issue:
Any vineyard ID greater than 2 results in the following error:
"auth_error_not_found": "Auth token not found."

scotte216 commented on src/admin/views.py:
```diff
@@ -0,0 +1,691 @@
+
+    try:
+        is_verified = verify_admin(admin_username, auth_token)
+        if (is_verified is False):
```
Could simplify these two lines
if not verify_admint(admin_username, auth_token): 

OR 

if not is_verified:

scotte216 commented:
General comment/observation:

We have the ability to disable a user and disable a vineyard. To enable a user you can use the /user/edit endpoint. According to the API documentation, the vineyard/edit endpoint doesn't seem to allow for re-enabling. 

Also, displaying user (admin/user) or vineyard information (admin/vineyard) doesn't show whether they are en-/disabled or not.

SoySauceLabs commented:
Sweet! Okay, so I've added a substantially more robust set of parameter validations and plenty of new unit tests. I believe I've addressed _all_ of the above comments. In addition, the API doc should be up to date with missing/invalid information!

scotte216 commented on src/admin/views.py:
```diff
@@ -29,22 +30,22 @@
-def verify_admin(admin_username, auth_token):
+def verify_admin(auth_token):
     """
     Verifies username has admin privileges and is authenticated.
```
Function description is inaccurate.

scotte216 commented:
It's not worth holding up the PR for a method description comment being inaccurate.

scotte216 commented:
So actually... When I create a new user and the subscription end-date has already passed, the log gets it right, but the returned error code is incorrect.

"unknown": "An unknown error occurred."
